### PR TITLE
[lite] feat: plug-and-play LoRA primitive (apply_lora_to_chunks)

### PR DESCRIPTION
**Pull request by ISEEKYAN**
Turn LoRA from an invasive build-time parameter into a post-build primitive,
mirroring how QAT is applied: the primitive owns attachment mechanics, while
native models contribute declarative target rules.

## What it does

- **`primitive/modules/lora_apply.py`** —
  `apply_lora_to_chunks(chunks, spec)` walks built model chunks, replaces
  supported target linears with LoRA-wrapped modules, and freezes base weights.
- **Structure-aware targeting** — attention, FFN, and MoE-expert surfaces are
  located by module contract. MLite uses TE and tensor-parallel linears, so
  matching only `torch.nn.Linear` would silently miss production modules.
- **No model-specific forward branches** — GQA and expert modules remain unaware
  of LoRA.
- **Orthogonal to QAT** — QAT parametrizes weights while LoRA wraps modules, so
  the two mechanisms compose.
- **Inert when disabled** — a disabled LoRA spec preserves the original model
  path and bitwise output.
- **Load-then-attach lifecycle** — canonical weights load before adapters are
  attached; attachment still occurs before FSDP2 or distributed-optimizer
  wrapping so optimizer buckets include adapter parameters.

## Usage

```python
chunks = protocol.build_model(config)
protocol.load_weights(chunks, checkpoint)
chunks = apply_lora_to_chunks(chunks, config.lora)
chunks = wrap_for_training(chunks)
```

The lifecycle order is part of the contract: loading after attachment would
insert wrapper-specific state-dict segments and break canonical checkpoint
mapping.

## Qwen3-30B-A3B contract

- 48 layers × 4 targets (`linear_qkv`, `linear_proj`, `linear_fc1`,
  `linear_fc2`) = **192 attached modules**.
- Rank 16 / alpha 32 with EP 8: **47,972,352** trainable adapter parameters
  out of **30,532,122,624** total (**0.1571209%**).
- Rank 128 / alpha 256 with EP 8: **383,778,816** trainable adapter parameters
  (**1.256967%**).

## Validation

- LoRA non-distributed CPU unit suites: **30 passed**.
- CPU/Gloo TP gradient suite: **3 passed**, including replicated grouped-expert
  adapter parity against a TP=1 reference.
- Real tiny-Qwen safetensors load through the production HF mapper:
  **0 missing keys, 0 unexpected keys** before adapter attachment.
- Initial parity: zero-initialized LoRA B matrices produce a
  **bitwise-identical** t0 output (`torch.equal`).
- Qwen3-30B-A3B structural and parameter-contract assertions cover both the
  192-module surface and trainable parameter counts.
- Compile and diff checks pass. The repository-pinned isort/Black hook pair
  still disagrees on existing whole-file formatting; its unrelated churn was removed
  rather than reported as a clean pre-commit run.
- Existing GPU evidence includes a LoRA-enabled Qwen3.5 SFT smoke with DDP and
  distributed optimizer.
- HF export no-grad regression coverage: **14 passed** in
  `test_hf_weights_streaming.py`.

## DAPO RL results

Five 30-step arms use the same 30B-A3B MoE model, DAPO math data, batch size 32,
response length 4096, 16 H100 GPUs, fused kernels, full actor offload, rollout
memory utilization 0.60, validation every five steps, checkpoint saving every
five steps, and AIME-2024 validation data. LoRA arms use rsLoRA; full
fine-tuning uses lr `1e-5`. All results are one run per arm.

The upstream 30B MoE recipe supplies lr `3e-6` and KL loss `0.001`; the upstream
best-practices guidance separately recommends rank 128 for a 32B-scale model.
The M1 arm combines those two recommendations with alpha 256.

| arm | adapter | lr | KL loss | validation acc@1 peak → final | reward median (steps 21–30) | peak allocated memory (GB) | median step time (s) |
|---|---|---:|---:|---:|---:|---:|---:|
| L3a | r32 / a64 | 1e-5 | off | 0.341 @30 → 0.341 | 0.117 | 39.27 | 102.7 |
| L4 | r128 / a256 | 1e-5 | off | 0.222 @10 → **0.000** | -0.945 | 40.78 | 105.8 |
| **M1 (upstream-aligned)** | **r128 / a256** | **3e-6** | **0.001** | **0.384 @20 → 0.363** | 0.180 | 39.58 | 120.8 |
| full-FT | none | 1e-5 | off | 0.415 @15 → 0.131 | 0.094 | 54.35 | 152.6 |
| **M2** | none | 1e-5 | **0.001** | **0.454 @10 → 0.372** | 0.250 | 51.94 | 159.9 |

### What the runs establish

- **The upstream-aligned rank-128 recipe is stable over this 30-step window.**
  M1 reaches 0.384 validation accuracy and finishes at 0.363. L4 learns
  initially, then degenerates: its final validation points are exactly zero,
  reward reaches exactly -1, and gradient norm reaches exactly zero.
- **Peak-then-decline is not LoRA-specific.** Full fine-tuning without KL peaks
  at 0.415 and falls to 0.131. In the clean full-FT pair, adding KL is the only
  configuration change and prevents that collapse in this run. This is an n=1
  observation, not a universal KL claim.
- **Parity is not established, and neither is a quality gap.** M1 and M2 finish
  at 0.363 and 0.372; their peaks differ by 0.070. An
  identical-configuration repeat elsewhere in this series moved validation by
  0.1375 despite the same seed. Both differences are below the experiment's
  measured single-run resolution. More steps in one run would not fix this;
  repeated runs per arm are required.
- **Resource savings are clear.** Across LoRA arms, peak allocated memory is
  38.7–40.8 GB versus 51.9–54.4 GB for full fine-tuning, a **24–29%**
  reduction. Median step time gives a **1.26–1.32×** LoRA speedup against the
  two full-FT controls.

Step-time claims use medians, not selected best steps. Observed min/max ranges
are 101.7–203.6 s for M1, 121.6–457.8 s for full-FT without KL, and
142.1–520.4 s for M2; comparing a best LoRA step with a worst full-FT step would
incorrectly inflate the speedup above 1.7×. Token-throughput ratios are not used
as the speed claim because the arms generate different token counts per step.

### Clean rank-128 resource pair on the final implementation

A final three-step pair exercised the pinned implementation commit `32753b4bd` with the same Qwen3-30B-A3B workload on one 8xH100 node per arm, for 16 GPUs running concurrently. Both arms used batch size 32, response length 4096, TP2/EP8/ETP1, fused kernels, parameter/gradient/optimizer offload, `optimizer_offload_fraction=1`, and rollout memory utilization 0.60. The LoRA arm used rank 128, alpha 128, and lr `1e-4`; the full-FT arm used lr `1e-5`.

- LoRA job **14606070**: **COMPLETED/0:0**, wall clock **40:10**.
- Full-FT job **14606073**: **COMPLETED/0:0**, wall clock **1:07:05**.
- LoRA reduced end-to-end elapsed time by **26:55**, or **40.1%**, corresponding to a **1.67x** wall-clock speedup.
- Logged `timing_s/step` values were 270.459/185.521/262.648 s for LoRA and 574.190/288.653/815.493 s for full-FT. Their three-step means are 239.543 s and 559.446 s, respectively.
- Peak actor allocation was **33.447 GiB** for LoRA versus **48.307 GiB** for full-FT, a **30.8%** reduction.

The three-step pair is resource evidence, not a statistical quality-parity claim. Its final validation accuracy was 0.071875 for LoRA and 0.079167 for full-FT.

## Scope and limitations

- All main-table arms used merged rollout weight synchronization. Their resource
  results therefore describe the merged-sync path.
- Train/rollout agreement metrics remain in the same range across the healthy
  arms, but they do not resolve quality differences below the measured
  validation noise floor.
- The experiment has one run per arm. A claim that LoRA matches—or fails to
  match—full fine-tuning requires repeated runs under the same frozen protocol.
- Response length is 4096 for all arms because of the current cross-entropy
  capacity limit; no 8192-token result is claimed.

## Tensor-parallel gradient validation

- For row-parallel `linear_proj`, a two-rank Gloo test compares the production
  LoRA path with a TP=1 full-matrix reference. Forward values, LoRA A/B
  gradients, and input gradients match within 1.8e-15; the LoRA A gradient-norm
  ratio is 1.0 on both ranks. Removing the backward reduction is a negative
  control and produces large gradient errors, confirming that the SUM combines
  distinct output-shard contributions rather than scaling a replicated
  gradient.

## Expert initialization caveat

- Expert adapters do not currently apply the PiSSA/OLoRA tail-initialization
  subtraction. With zero-initialized `lora_b`, the initial adapter delta is
  zero, so skipping the subtraction is numerically equivalent to subtracting
  zero at t0. Non-zero expert tail initialization remains unsupported; the
  initializer now emits a `UserWarning` whenever grouped expert adapters are
  skipped while continuing to initialize supported dense adapters.

## QKV tensor/sequence-parallel normalization

- All DAPO LoRA arms used actor TP=2, which enables sequence parallelism, and
  targeted `linear_qkv`; the full-FT controls did not enter the LoRA wrapper.
- Transformer Engine normalizes each `[S/TP, B, H]` shard over the final hidden
  dimension before gathering along the sequence dimension for the
  column-parallel GEMM. A controlled TP=2 proxy found a **0.0** maximum
  difference between shard-local RMSNorm followed by sequence concatenation
  and the full-sequence tokenwise reference; sequence parallelism did not
  invalidate the five-arm results.
- The implementation no longer recomputes RMSNorm. It enables Transformer
  Engine `return_layernorm_output` and feeds that exact kernel output to the QKV
  adapter. A focused regression asserts that the adapter receives the exact
  tensor object returned by the base normalized-input surface. A non-zero
  TP=2/SP Gloo regression also checks the adapter input elementwise and verifies
  training forward, materialized delta, and `(W+BA)` merged export within
  **1e-12**.

## Native-model and configuration coverage

- Qwen3-MoE, Qwen3.5, GLM5, Kimi K2, and DeepSeek V4 now provide
  declarative LoRA target rules. The shared primitive contains no model-name
  branches.
- Real tiny-model regressions instantiate each native model, verify that every
  declared target is attached exactly once, check exact trainable tensor and scalar counts and parameter ownership,
  and execute the adapter forward path.
- `ignore_patterns` is enforced on both common and model-declared attachment
  paths using case-insensitive dotted-path components. Regressions verify that
  matched modules remain unwrapped.
- The independent model-by-capability matrix (attach, TP gradients, merged
  export, PEFT round-trip, and ignore filtering) is maintained in
  `experimental/lite/examples/verl/LoRA.md`; unimplemented cells are marked
  explicitly rather than inferred from nearby tests.

## Additional contract caveats

- Expert tensor parallelism is outside the supported LoRA scope.
  `apply_lora_to_chunks` rejects ETP>1 before mutating the model, so unsupported
  training and PEFT round-trips cannot silently proceed.
- `enabled: true` remains the only LoRA opt-in. A positive `rank` with LoRA
  disabled stays inert but now emits a `UserWarning` explaining the required
  opt-in.
- `olora_tail` does not cover grouped MoE expert adapters. Their
  zero-initialized B matrices make the initial delta zero; the initializer emits
  one `UserWarning` per call when grouped adapters are skipped, and non-zero
  expert tail initialization remains unsupported.
- Dedicated PEFT round-trip coverage remains limited to the Qwen3-MoE mapping;
  the coverage matrix marks the other native models as unimplemented rather
  than claiming support from attachment tests alone.

### CODE REVIEW

**Comment by ISEEKYAN** [comment]
## Formal rebuttal: TP>1 QKV LoRA normalization

**Requested disposition: withdraw the blocker as stated.** The path is real, but the causal step “SP-local RMSNorm differs from TE sequence-parallel RMSNorm” is false for this Qwen3-MoE layout.

### Causal-chain audit

1. **`ACTOR_TP=2` and QKV targeting: confirmed.** The resolved five-arm contract uses actor TP=2. Every LoRA arm targets `linear_qkv`; full-FT does not enter the LoRA wrapper.
2. **Sequence parallelism: confirmed.** Qwen3-MoE scatters activations along the sequence dimension (`model/qwen3_moe/lite/model.py:250,260`). `ColumnParallelLinear` enables `use_sp` for TP>1 (`primitive/parallel/linear.py:186-199`). The local tensor is `[S/TP, B, H]`; the hidden dimension `H` remains complete on every TP rank.
3. **“Local RMSNorm differs from TE SP-aware normalization”: refuted.** RMSNorm reduces the final hidden dimension, not the sequence dimension. NVIDIA Transformer Engine commit [`6d3280c8`](https://github.com/NVIDIA/TransformerEngine/blob/6d3280c8255e6f02bb6af282447eaafe68d04a87/transformer_engine/pytorch/module/layernorm_linear.py#L248-L306) calls `apply_normalization(inputmat, ...)` first (lines 248–261), then all-gathers `ln_out` along the first dimension (lines 276–306). A controlled TP=2/SP proxy comparing shard-local RMSNorm followed by sequence concatenation against the full-sequence tokenwise reference produced **max_abs_diff = 0.0**.
4. **Training vs merged rollout divergence: does not follow.** With the same normalized input `N(x)`, `W N(x) + BA N(x) = (W+BA) N(x)`. The non-zero TP=2/SP regression verifies adapter input identity, training forward, materialized delta, and merged-export forward within **1e-12**.

### Hardening despite the rebuttal

The rejected SP premise does not prove that independently recomputing a fused kernel output is a good contract. Commit `e5664b93c` therefore removes the handwritten QKV RMSNorm and feeds the adapter Transformer Engine’s exact `return_layernorm_output`. Commit `64b69b56c` adds a fail-loud guard: any normalized QKV surface that cannot expose its exact normalized tensor is rejected rather than silently recomputed. The LoRA primitive suite is **16 passed**, including the existing row-parallel TP gradient parity test.

### Boundary where this class of blocker would be valid

The mismatch would be real if any of these conditions held:

- the normalization reduction axis itself were sharded (for example, QKV input carried only `H/TP`, requiring a cross-rank reduction);
- base and adapter used different epsilon, gamma/zero-centered-gamma semantics, dtype/quantization, or different pre/post-normalization tensors;
- a gathered normalized tensor were gathered again, or a normalized QKV surface silently fell back to raw `x`.

None is reachable in the current Qwen3-MoE GQA path: SP shards `S`, not `H`, and the adapter now consumes the base kernel output directly. A future normalized surface without that protocol hits the new guard. TP=1 and target sets excluding `linear_qkv` are unaffected by construction.

### Experimental disposition

The five-arm table does **not** require reclassification or rerun because of this blocker. Its validation, memory, and step-time observations remain admissible. The table is already in this PR and all included runs use `test_freq=5`, `save_freq=5`, and AIME-2024 validation data. The existing statistical limitation remains: one run per arm and a measured 0.1375 repeat-run validation spread mean the data neither proves nor disproves LoRA/full-FT quality parity; it does support the reported resource differences.

**Comment by ISEEKYAN** [comment]
### TP/SP QKV normalization concern: resolved

The current implementation does not recompute normalization in the LoRA branch:

- [`LoRAWrappedLinear.forward`](https://github.com/ISEEKYAN/Megatron-LM/blob/c28f5409be5d22c4db661fe1909810e38318be49/experimental/lite/megatron/lite/primitive/modules/lora_apply.py#L47-L55) obtains `(out, adapter_x)` from `base.forward_with_normalized_input(...)` and passes that exact `adapter_x` object to the adapter.
- [`_enable_qkv_normalized_output`](https://github.com/ISEEKYAN/Megatron-LM/blob/c28f5409be5d22c4db661fe1909810e38318be49/experimental/lite/megatron/lite/primitive/modules/lora_apply.py#L123-L133) enables Transformer Engine’s `return_layernorm_output` surface; it does not implement normalization itself.
- A full-file search of `lora_apply.py` for `rsqrt`, `variance`, `pow`, and `mean` returns zero matches.

Therefore, the adapter consumes Transformer Engine’s own layernorm output by construction, including when sequence parallelism is enabled. The proposed mismatch path does not exist.

Commit `c28f5409b` adds an [object-identity regression](https://github.com/ISEEKYAN/Megatron-LM/blob/c28f5409be5d22c4db661fe1909810e38318be49/experimental/lite/tests/unit/primitive/test_apply_lora_to_chunks_unit.py#L412-L430) asserting that the adapter receives the exact tensor object returned by the base normalized-input surface, in addition to the existing TP=2/SP elementwise and merged-export regression. It also makes the separate `olora_tail` grouped-expert limitation non-silent: supported dense adapters are initialized, while skipped MoE grouped adapters emit a `UserWarning` and remain explicitly unsupported.

**Comment by ISEEKYAN** [comment]
CR-01 is confirmed and fixed in f856a9b19.

The new `test_shared_grouped_tp2_gradients_match_tp1` reproduces the issue with two CPU/Gloo TP ranks: before the fix, the replicated `lora_a` gradients differed in all 18 elements across ranks (maximum absolute difference 10.468008390582618). `lora_b` was covered by the same parity contract.

The fix threads the TP process group into grouped expert adapters only for `tp > 1, ep = 1, etp = 1`, then SUM-reduces the replicated LoRA parameter gradients with the existing `_all_reduce_sum` primitive. The test now verifies that both TP2 ranks have identical `lora_a`/`lora_b` gradients and that each matches the TP1 reference at `atol=1e-12`.

Validation after the final diff cleanup:
- `test_export_merge_lora_unit.py`
- `test_apply_lora_to_chunks_unit.py`
- `test_lora_tp_gradients.py`

Result: 21 passed, 14 warnings. `git diff --check` also passes.

**Comment by ISEEKYAN** [comment]
CR-01 final disposition: resolved by the implementation, not rebutted as a false positive.

The pre-fix negative control showed that TP=2 replicated grouped-expert LoRA gradients were incorrect: all 18 `lora_a` elements differed across ranks, with a maximum absolute difference of 10.468008390582618. The TP SUM gradient hook therefore remains required.

On the current PR head (`32753b4bd`), the focused two-rank CPU/Gloo regression was rerun:

`GLOO_SOCKET_IFNAME=lo PYTHONPATH=experimental/lite pytest -q experimental/lite/tests/unit/primitive/test_lora_tp_gradients.py::test_shared_grouped_tp2_gradients_match_tp1`

Result: **1 passed in 12.08s**, exit code 0. The test requires both TP=2 ranks to have identical `lora_a` and `lora_b` gradients and requires each rank to match the TP=1 reference at `atol=1e-12`. This closes CR-01 as fixed and regression-locked.
